Generates invalid C when omitting self in lambda

I'm using Cython version 0.29.7.

Part of the error message (full message attached: [output.txt](https://github.com/cython/cython/files/3207113/output.txt)):
```
AA/bug.c:830:57: error: expected identifier or ‘(’ before ‘.’ token
 struct __pyx_obj_2AA_3bug___pyx_scope_struct____pyx_base.func;
```

Here's a quick example to reproduce, including a workaround.

```
# cython: language_level=3
cdef class FBase():
    def __init__(self):
        pass
        
    def __cinit__(self):
        pass
        
    cdef func(self, whatever):
        pass
        

cdef class Bug(FBase):
    def __init__(self):
        pass
        
    cdef check(self, x):
        return x
        
    cdef func(self, whatever):
        a = [1, 2, 3]
        min(a, key=lambda x: self.check(x))

        # This generates correct C, but doesn't work in Python:
        # min(a, key=lambda self, x: self.check(x))

        # Workaround:
        # import functools
        # min(a, key=functools.partial(lambda self, x: self.check(x), self))
```

--- a/Cython/Compiler/ParseTreeTransforms.py
+++ b/Cython/Compiler/ParseTreeTransforms.py
@@ -2738,9 +2738,12 @@ class CreateClosureClasses(CythonTransform):
             node.needs_outer_scope = True
             return
 
+        # entry.cname can contain periods (eg. a derived C method of a class).
+        # We want to use the cname as part of a C struct name, so we replace
+        # periods with double underscores.
         as_name = '%s_%s' % (
             target_module_scope.next_id(Naming.closure_class_prefix),
-            node.entry.cname)
+            node.entry.cname.replace('.','__'))
 
         entry = target_module_scope.declare_c_class(
             name=as_name, pos=node.pos, defining=True,

--- a/tests/run/closure_in_derived_class_T2967.pyx
+++ b/tests/run/closure_in_derived_class_T2967.pyx
@@ -1,0 +1,18 @@
+# mode: run
+# tag: closures
+# ticket: 2967
+
+cdef class BaseClass:
+    cdef func(self):
+        pass
+cdef class ClosureInsideExtensionClass(BaseClass):
+    """
+    >>> y = ClosureInsideExtensionClass(42)
+    >>> y.test(42)
+    43
+    """
+    cdef func(self):
+        a = 1
+        return (lambda x : x+a)
+    def test(self, b):
+        return self.func()(b)

